Source/WebCore/page/Navigation.cpp

```diff
@@ -613,6 +614,50 @@ void Navigation::abortOngoingNavigation(NavigateEvent& event)
     }
 }
 
+static std::pair<Ref<DOMPromise>, Ref<DeferredPromise>> createPromiseAndWrapper(Document& document)
+{
+    auto& globalObject = *JSC::jsCast<JSDOMGlobalObject*>(document.globalObject());
+    JSC::JSLockHolder lock(globalObject.vm());
+    RefPtr deferredPromise = DeferredPromise::create(globalObject);
+    Ref domPromise = DOMPromise::create(globalObject, *JSC::jsCast<JSC::JSPromise*>(deferredPromise->promise()));
+    return { WTFMove(domPromise), deferredPromise.releaseNonNull() };
+}
+
+// https://webidl.spec.whatwg.org/#wait-for-all
+static void waitForAllPromises(Vector<RefPtr<DOMPromise>>& promises, std::function<void()> fulfilledCallback, std::function<void(JSC::JSValue)> rejectionCallback)
+{
+    auto remainingPromises = std::make_shared<size_t>(promises.size());
```

Comment:
WebKit usually uses RefPtr and RefCounted instead of std::shared_ptr.  Can we make a structure of the state?  Can we use CallbackAggregator instead?